src/ray/util/logging.cc

```diff
@@ -418,58 +499,84 @@ void RayLog::AddFatalLogCallbacks(
+                                              line_number,
+                                              pid,
+                                              std::to_string(GetTid()),
+                                              errno,
+                                              strerror(errno));
   }
   if (is_enabled_) {
-    logging_provider_ = new LoggingProvider(
-        file_name, line_number, GetMappedSeverity(severity), expose_osstream_);
+    if (log_format_json_) {
+      if (!component_name_.empty()) {
+        context_osstream_ << ",\"" << kLogKeyComponent << "\":\"" << component_name_
+                          << "\"";
+      }
+      context_osstream_ << ",\"" << kLogKeyFilename << "\":\"" << ConstBasename(file_name)
+                        << "\",\"" << kLogKeyLineno << "\":" << line_number;
+    } else {
+      if (!component_name_.empty()) {
+        msg_osstream_ << "(" << component_name_ << ") ";
+      }
+      msg_osstream_ << ConstBasename(file_name) << ":" << line_number << ": ";
+    }
   }
 }
 
-std::ostream &RayLog::Stream() {
-  auto logging_provider = reinterpret_cast<LoggingProvider *>(logging_provider_);
-  // Before calling this function, user should check IsEnabled.
-  // When IsEnabled == false, logging_provider_ will be empty.
-  return logging_provider->stream();
-}
-
 bool RayLog::IsEnabled() const { return is_enabled_; }
 
 bool RayLog::IsFatal() const { return is_fatal_; }
 
-std::ostream &RayLog::ExposeStream() { return *expose_osstream_; }
-
 RayLog::~RayLog() {
-  if (logging_provider_ != nullptr) {
-    delete reinterpret_cast<LoggingProvider *>(logging_provider_);
-    logging_provider_ = nullptr;
-  }
-  if (expose_osstream_ != nullptr) {
+  if (IsFatal()) {
+    msg_osstream_ << "\n*** StackTrace Information ***\n" << ray::StackTrace();
+    expose_fatal_osstream_ << "\n*** StackTrace Information ***\n" << ray::StackTrace();
     for (const auto &callback : fatal_log_callbacks_) {
-      callback(EL_RAY_FATAL_CHECK_FAILED, expose_osstream_->str());
+      callback(EL_RAY_FATAL_CHECK_FAILED, expose_fatal_osstream_.str());
     }
   }
+
+  auto logger = spdlog::get(RayLog::GetLoggerName());
+  if (!logger) {
+    logger = DefaultStdErrLogger::Instance().GetDefaultLogger();
+  }
+  // NOTE(lingxuan.zlx): See more fmt by visiting https://github.com/fmtlib/fmt.
+  if (log_format_json_) {
+    logger->log(GetMappedSeverity(severity_),
+                /*fmt*/ ",\"{}\":\"{}\"{}",
+                kLogKeyMessage,
+                json_escape_string(msg_osstream_.str()),
+                context_osstream_.str());
+  } else {
+    logger->log(GetMappedSeverity(severity_),
+                /*fmt*/ "{}{}",
+                msg_osstream_.str(),
+                context_osstream_.str());
+  }
+  logger->flush();
+
   if (severity_ == RayLogLevel::FATAL) {
     std::_Exit(EXIT_FAILURE);
   }
 }
 
+RayLog &RayLog::WithField(const std::string &key, const std::string &value) {
```

Comment:
This API requires all fields happen before any text strings, for which I am fine, though it *may* introduce some inconveniences when we have some optional outputs.

However it also only accepts strings, which seems to be a big deal, since it forces a `base_id.Hex()` call. What about:

```
template <typename T>
RayLog &RayLog::WithField(std::string_view key, const T &value) {
    std::stringstream ss;
    ss << value;
    // put ss.str() to field_ or context_osstream_...
}
```